reconcile/openshift_base.py

```diff
@@ -556,6 +559,82 @@ def check_unused_resource_types(ri):
             logging.warning(msg)
 
 
+def _normalize_secret(secret: OR):
+    body = secret.body
+    string_data = body.pop("stringData", None)
+    if string_data:
+        body.setdefault("data", {})
+        for k, v in string_data.items():
+            v = base64.b64encode(str(v).encode()).decode("utf-8")
+            body["data"][k] = v
+
+
+def _normalize_noop(item: OR):
+    pass
+
+
+def normalize_object(item: OR, deep_copy=True) -> OR:
+    if deep_copy:
+        n = copy.deepcopy(item)
```

Comment:
Can we avoid deepcopy by cherry pick fields to create new object, instead of mutate in place?

Reply:
Yes, changed the method to create a new object instead of copying it.

Reply:
There is one issue with that new `OR` object, `body=item.body` so `item.body` will pass as ref, then any mutation to `n.body` will also affect original body, is this expected?

An easy way to use without worry about mutation is always return a new object.
For example:

To add or override a dict entry, instead of `n.body["metadata"] = metadata`, can use `n.body = n.body | {"metadata": metadata}`.

To remove a dict entry, instead of `annotations.pop("qontract.integration", None)`, can use `new_annotations = {k: v for k, v in body.items() if k != "qontract.integration"}`

Reply:
that only solved one layer of copy, nested dict object is still a reference to original one, so it still mutate original object, to make it fully decoupled, there are a few changes needed:

```python
def _normalize_secret(secret: OR) -> None:
    body = secret.body
    string_data = body.get("stringData")
    if string_data:
        data = body.get("data") or {}
        body["data"] = data | {
            k: base64.b64encode(str(v).encode()).decode("utf-8")
            for k, v in string_data.items()
        }
        secret.body = {k: v for k, v in body.items() if k != "stringData"}

QONTRACT_ANNOTATIONS = {
    "qontract.integration",
    "qontract.integration_version",
    "qontract.sha256sum",
    "qontract.update",
    "qontract.caller_name",
}

def normalize_object(item: OR) -> OR:
    # Remove K8s managed attributes not needed to compare objects
    metadata = {
        k: v
        for k, v in item.body["metadata"].items()
        if k not in NORMALIZE_COMPARE_EXCLUDED_ATTRS
    }

    annotations = metadata.get("annotations") or {}

    metadata["annotations"] = {
        k: v for k, v in annotations.items() if k not in QONTRACT_ANNOTATIONS
    }

    n = OR(
        body=item.body | {"metadata": metadata},
        integration=item.integration,
        integration_version=item.integration_version,
        error_details=item.error_details,
        caller_name=item.caller_name,
        validate_k8s_object=False,
    )

    # Run normalizers on Kinds with special needs
    NORMALIZERS.get(n.body["kind"], _normalize_noop)(n)
    return n
```

In this way, for every dict operation, only root level field assign value, without any side effect of original dict.

Reply:
Fixed. PTAL

Reply:
Nice! https://github.com/app-sre/qontract-reconcile/pull/3768/files#diff-e1ae1861e6058b02d61980281bb71def45745bdfc82c269bfcf2a52f6be4fabdR295 just left for future usage, right? Not used noe.

Reply:
Method Removed.